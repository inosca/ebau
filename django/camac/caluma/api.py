from logging import getLogger

from caluma.caluma_form import models as caluma_form_models
from caluma.caluma_form.validators import AnswerValidator, DocumentValidator
from caluma.caluma_workflow import (
    api as caluma_workflow_api,
    models as caluma_workflow_models,
)
from django.conf import settings
from django.db.models import Q

from camac.user.models import Service

log = getLogger(__name__)


class CalumaApi:
    """
    Class with helper methods to interact with Caluma.

    Initially this was meant as a wrapper for all interactions with Caluma (including
    the models). This is not the case anymore. Instead this class is meant to contain
    convenience methods for more involved caluma interactions or project specific
    meta lookups.
    """

    def _get_main_form(self, instance):
        return instance.case.document.form if instance.case else None

    def get_form_name(self, instance):
        form = self._get_main_form(instance)
        return form.name if form else None

    def get_form_slug(self, instance):
        form = self._get_main_form(instance)
        return form.slug if form else None

    def _get_main_workflow(self, instance):
        return instance.case.workflow if instance.case else None

    def get_workflow_slug(self, instance):
        workflow = self._get_main_workflow(instance)
        return workflow.slug if workflow else None

    def _get_main_document(self, instance):
        return instance.case.document

    def get_main_document(self, instance):
        document = self._get_main_document(instance)
        return document.pk if document else None

    def get_source_document_value(self, document_id, field):
        source = caluma_form_models.Document.objects.get(pk=document_id).source
        return getattr(source, field, None) if source else None

    def get_ebau_number(self, instance):
        return instance.case.meta.get("ebau-number", "-")

    def get_dossier_number(self, instance):
        return instance.case.meta.get("dossier-number", "-")

    def get_municipality(self, instance):
        return self.get_answer_value("gemeinde", instance)

    def get_answer_value(self, question_slug, instance):
        answer = instance.case.document.answers.filter(
            question_id=question_slug
        ).first()

        return answer.value if answer else None

    def get_table_answer(self, question_slug, instance):
        try:
            answer = instance.case.document.answers.get(
                question_id=question_slug,
            )
            return answer.documents.all()
        except caluma_form_models.Answer.DoesNotExist:
            return None

    def get_nfd_form_permissions(self, instance):
        permissions = set()

        answers = caluma_form_models.Answer.objects.filter(
            question_id="nfd-tabelle-status",
            document__family__form_id="nfd",
            document__family__work_item__case__family__instance__pk=instance.pk,
        )

        if answers.exclude(value="nfd-tabelle-status-entwurf").exists():
            permissions.add("read")

        if answers.filter(value="nfd-tabelle-status-in-bearbeitung").exists():
            permissions.add("read")
            permissions.add("write")

        return permissions

    def copy_document(self, source_pk, exclude_form_slugs=None, meta=None, **kwargs):
        """Use to `copy()` function on a document and do some clean-up.

        Caution: `exclude_form_slugs` is only excluding top-level questions
        and doesn't do additional clean-up on nested documents from
        table questions. That's based on the assumption that there are no table
        questions in the excluded form.
        """
        source = caluma_form_models.Document.objects.get(pk=source_pk)
        document = source.copy(**kwargs)

        if exclude_form_slugs:
            document.answers.filter(question__forms__in=exclude_form_slugs).delete()

        # prevent creating a historical record
        document.skip_history_when_saving = True
        try:
            document.save()
        finally:
            del document.skip_history_when_saving

        return document

    def update_or_create_answer(self, document, question_slug, value, user):
        question = caluma_form_models.Question.objects.get(slug=question_slug)

        AnswerValidator().validate(
            question=question, document=document, user=user, value=value
        )

        return caluma_form_models.Answer.objects.update_or_create(
            document=document,
            question_id=question_slug,
            defaults={"value": value},
        )

    def set_submit_date(self, instance_id, submit_date):
        case = caluma_workflow_models.Case.objects.get(instance__pk=instance_id)

        if "submit-date" in case.meta:  # pragma: no cover
            # instance was already submitted, this is probably a re-submit
            # after correction.
            return False

        new_meta = {
            **case.meta,
            # Caluma date is formatted yyyy-mm-dd so it can be sorted
            "submit-date": submit_date,
        }

        case.meta = new_meta
        case.save()

        return True

    def is_paper(self, instance):
        return instance.case.document.answers.filter(
            question_id="is-paper",
            value="is-paper-yes",
        ).exists()

    def is_modification(self, instance):
        return instance.case.document.answers.filter(
            question_id="projektaenderung",
            value="projektaenderung-ja",
        ).exists()

    def is_migrated(self, instance):
        return instance.case.workflow_id == "migrated"

    def get_migration_type(self, instance):
        answer = instance.case.document.answers.filter(
            question_id="geschaeftstyp"
        ).first()

        if not answer:  # pragma: no cover
            return None

        option = answer.question.options.get(slug=answer.value)

        return (option.slug, option.label)

    def copy_table_answer(
        self,
        source_question,
        target_question,
        source_document,
        target_document,
        source_question_fallback=None,
    ):
        # get the source answer with a fallback
        table_answer = source_document.answers.filter(
            question_id=source_question
        ).first()

        if not table_answer or table_answer.documents.count() == 0:
            table_answer = source_document.answers.filter(
                question_id=source_question_fallback
            ).first()

        # nothing to copy
        if not table_answer or table_answer.documents.count() == 0:  # pragma: no cover
            return

        # create a new table answer in the target document
        new_table_answer = caluma_form_models.Answer.objects.create(
            document_id=target_document.pk, question_id=target_question
        )

        # copy all rows into the new table answer
        for row in table_answer.documents.all():
            sb_row = self.copy_document(row.id, family=target_document.family)
            new_table_answer.documents.add(sb_row)

    def reassign_work_items(self, instance, from_group_id, to_group_id, user):
        from_group_id = str(from_group_id)
        to_group_id = str(to_group_id)

        for work_item in (
            caluma_workflow_models.WorkItem.objects.filter(
                Q(addressed_groups__contains=[from_group_id])
                | Q(controlling_groups__contains=[from_group_id])
            )
            .filter(
                status__in=[
                    caluma_workflow_models.WorkItem.STATUS_READY,
                    caluma_workflow_models.WorkItem.STATUS_SUSPENDED,
                ],
                case__family__instance=instance,
            )
            .exclude(
                task_id__in=[
                    "create-manual-workitems",
                    settings.DISTRIBUTION["INQUIRY_TASK"],
                    settings.DISTRIBUTION["INQUIRY_ANSWER_FILL_TASK"],
                    settings.DISTRIBUTION["INQUIRY_CREATE_TASK"],
                    settings.DISTRIBUTION["INQUIRY_CHECK_TASK"],
                ]
            )
        ):
            for groups_type in ["addressed_groups", "controlling_groups"]:
                groups = set(getattr(work_item, groups_type))

                if from_group_id not in groups:
                    continue

                groups.remove(from_group_id)
                groups.add(to_group_id)

                # If the addressed groups change, we need to filter out all
                # assigned users that are not member of the new addressed group
                if len(work_item.assigned_users) and groups_type == "addressed_groups":
                    work_item.assigned_users = [
                        username
                        for username in work_item.assigned_users
                        if Service.objects.filter(
                            pk=int(to_group_id), groups__users__username=username
                        ).exists()
                    ]

                setattr(work_item, groups_type, list(groups))

            work_item.save()

        # If there is no work item to allow creation of an inquiry for the new
        # service and the distribution is still running, we need to create one
        distribution = instance.case.work_items.filter(
            task_id=settings.DISTRIBUTION["DISTRIBUTION_TASK"],
            status=caluma_workflow_models.WorkItem.STATUS_READY,
        ).first()

        if (
            distribution
            and not distribution.child_case.work_items.filter(
                task_id=settings.DISTRIBUTION["INQUIRY_CREATE_TASK"],
                addressed_groups__contains=[to_group_id],
                status=caluma_workflow_models.WorkItem.STATUS_READY,
            ).exists()
        ):
            task = caluma_workflow_models.Task.objects.get(
                pk=settings.DISTRIBUTION["INQUIRY_CREATE_TASK"]
            )

            caluma_workflow_models.WorkItem.objects.create(
                task=task,
                name=task.name,
                addressed_groups=[to_group_id],
                controlling_groups=[to_group_id],
                case=distribution.child_case,
                status=caluma_workflow_models.WorkItem.STATUS_READY,
                created_by_user=user.username,
                created_by_group=user.group,
            )

    def validate_existing_audit_documents(self, instance_id, user):
        """Intermediate validation of existing audits.

        Make sure that those documents which are linked to the three
        table questions covering the "audit" functionality are valid.

        This is used when the responsible service is changed, where we
        want to make sure that filled audits are valid, but the entire
        document doesn't have to be valid yet.
        """
        caluma_settings = settings.APPLICATION.get("CALUMA", {})

        audit_work_item = caluma_workflow_models.WorkItem.objects.filter(
            **{
                "task_id": caluma_settings.get("AUDIT_TASK"),
                "status": caluma_workflow_models.WorkItem.STATUS_READY,
                "case__family__instance__pk": instance_id,
            }
        ).first()

        if not audit_work_item:
            return

        for table_answer in audit_work_item.document.answers.filter(
            question__type=caluma_form_models.Question.TYPE_TABLE
        ):
            for document in table_answer.documents.all():
                DocumentValidator().validate(document, user)

    def close_publication(self, instance, user):
        caluma_settings = settings.APPLICATION.get("CALUMA", {})
        publication_slug = caluma_settings.get("PUBLICATION_TASK_SLUG", "")

        work_item = caluma_workflow_models.WorkItem.objects.filter(
            status=caluma_workflow_models.WorkItem.STATUS_READY,
            case__family__instance__pk=instance.pk,
            task__slug=publication_slug,
        ).first()

        if work_item:
            caluma_workflow_api.complete_work_item(work_item, user)

    def _get_answer(self, document, slug):
        return str(
            document.answers.filter(question_id=slug)
            .values_list("value", flat=True)
            .first()
            or ""
        ).strip()

    def get_address(self, document):
        street, number, city, migrated = [
            self._get_answer(document, ans)
            for ans in [
                "strasse-flurname",
                "nr",
                "ort-grundstueck",
                "standort-migriert",
            ]
        ]

        address_lines = [f"{street} {number}".strip(), city]

        return ", ".join(address_lines) if all(address_lines) else migrated

    def get_gesuchsteller(self, document):
        table_ans = document.answers.filter(
            question_id="personalien-gesuchstellerin"
        ).first()
        if not table_ans or not table_ans.documents.exists():  # pragma: no cover
            return ""

        def _get_name(doc):
            name_jurist, first_name, last_name = [
                self._get_answer(doc, ans)
                for ans in [
                    "name-juristische-person-gesuchstellerin",
                    "vorname-gesuchstellerin",
                    "name-gesuchstellerin",
                ]
            ]

            return name_jurist or f"{first_name} {last_name}".strip()

        return ", ".join(  # pragma: no cover
            [_get_name(row_doc) for row_doc in table_ans.documents.all()]
        )

    def get_gemeinde(self, document):
        return self._get_answer(document, "gemeinde")
