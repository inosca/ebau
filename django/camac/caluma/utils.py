from copy import copy
from typing import Optional

from caluma.caluma_form.models import Answer, Document, Question
from caluma.caluma_user.models import AnonymousUser, OIDCUser
from django.conf import settings
from django.contrib.auth.models import AnonymousUser as AnonymousCamacUser
from django.utils.translation import get_language
from jwt import decode as jwt_decode
from rest_framework.authentication import get_authorization_header

from camac.user.models import User
from camac.user.utils import get_group


def extend_user(user, camac_request):
    """Patch the caluma user to contain the needed data.

    This will set the caluma group (which is used in created_by_group etc.)
    to the CAMAC service (!). It will also set the properties `camac_role`
    and `camac_group` in case they are needed.
    """

    if camac_request.group:
        user.camac_role = camac_request.group.role.name
        user.camac_group = camac_request.group.pk
        user.group = camac_request.group.service_id

    return user


def find_answer(document: Document, question: str, **kwargs) -> str:
    """
    Find the answer to a certain question in a document.

    >>> find_answer(
    ...     document=Document.objects.first(),
    ...     question="my-question"
    ... )
    'The answer to my question'
    """
    answer = (
        document.answers.select_related("question")
        .prefetch_related("question__options")
        .filter(question_id=question)
        .first()
    )

    if not answer:
        return ""

    return get_answer_display_value(answer, **kwargs)


def get_answer_display_value(
    answer: Answer,
    option_separator: Optional[str] = ", ",
    date_format: Optional[str] = settings.MERGE_DATE_FORMAT,
    language: Optional[str] = get_language(),
) -> str:
    """
    Get the display value of an answer depending on the question type.

    >>> get_answer_display_value(
    ...     answer=Answer.objects.get(question_id="date-question")
    ... )
    '02.06.2022'
    """
    if answer.question.type in [Question.TYPE_MULTIPLE_CHOICE, Question.TYPE_CHOICE]:
        return option_separator.join(
            [option.label.get(language) for option in answer.selected_options]
        )
    elif answer.question.type == Question.TYPE_DATE:
        return answer.date.strftime(date_format)

    return answer.value


class CamacRequest:
    """
    A camac request object built from the given caluma info object.

    The request attribute holds a shallow copy of `info.context` with translated
    values where needed (user, group, etc.).
    """

    def __init__(self, info):
        self.request = copy(info.context)
        oidc_user = self.request.user
        self.request.user = self._get_camac_user(oidc_user)
        self.request.auth = (
            jwt_decode(oidc_user.token, options={"verify_signature": False})
            if oidc_user.token
            else None
        )
        camac_group = get_group(self.request)
        self.request.group = camac_group
        self.request.oidc_user = oidc_user
        self.request.query_params = self.request.GET

    def _get_camac_user(self, oidc_user):
        if isinstance(oidc_user, AnonymousUser):
            return AnonymousCamacUser()

        return User.objects.get(username=oidc_user.username)


class CalumaInfo:
    """A caluma info object built from the given camac request.

    Caluma requires an "info" object in various places, representing
    the GraphQL request, user, etc; similar to the context in
    DRF views.

    This info object is limited and only contains what's actually needed.
    It may need to be expanded in the future.
    """

    def __init__(self, request):
        self.context = CalumaInfo._Context(request)

    class _Context:
        def __init__(self, request):
            _, token = get_authorization_header(request).split()
            oidc_user = OIDCUser(token=token, userinfo=request.auth)

            self.user = extend_user(oidc_user, request)
