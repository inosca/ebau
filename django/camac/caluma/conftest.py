import functools
from collections import namedtuple

import pytest
from caluma.schema import schema

from camac.caluma.utils import CamacRequest


@pytest.fixture
def caluma_admin_request(rf, caluma_admin_user):
    def wrapper(**kwargs):
        request = rf.get("/graphql", **kwargs)
        request.user = caluma_admin_user
        Info = namedtuple("Info", "context")
        request.camac_request = CamacRequest(Info(context=request)).request
        return request

    return wrapper


@pytest.fixture
def caluma_admin_schema_executor(caluma_admin_request):
    return functools.partial(
        schema.execute, context_value=caluma_admin_request(), middleware=[]
    )


@pytest.fixture
def caluma_admin_public_schema_executor(caluma_admin_request):
    return functools.partial(
        schema.execute,
        context_value=caluma_admin_request(HTTP_X_CAMAC_PUBLIC_ACCESS=True),
        middleware=[],
    )
