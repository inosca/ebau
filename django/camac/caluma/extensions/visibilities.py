from caluma.caluma_core.visibilities import filter_queryset_for
from caluma.caluma_form import (
    historical_schema as historical_form_schema,
    schema as form_schema,
)
from caluma.caluma_user.visibilities import Authenticated
from caluma.caluma_workflow import schema as workflow_schema
from django.conf import settings
from django.contrib.postgres.fields import ArrayField
from django.db.models import Case, Exists, F, Func, IntegerField, OuterRef, Q, When
from django.db.models.functions import Cast

from camac.caluma.utils import CamacRequest
from camac.constants.kt_bern import DASHBOARD_FORM_SLUG
from camac.instance.filters import CalumaInstanceFilterSet
from camac.instance.mixins import InstanceQuerysetMixin
from camac.user.models import Role, Service
from camac.utils import filters, order


class CustomVisibility(Authenticated, InstanceQuerysetMixin):
    """Custom visibility for Kanton Bern.

    To avoid multiple db lookups, the result is cached in the
    request object, and reused if the need arises. Caching beyond a request is
    not done but might become a future optimisation.
    """

    instance_field = None

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.group = None
        self.user = None
        self.request = None

    def filter_queryset(self, node, queryset, info):
        if not getattr(info.context, "camac_request", None):
            info.context.camac_request = CamacRequest(info).request

        self.request = info.context.camac_request

        return super().filter_queryset(node, queryset, info)

    @filter_queryset_for(form_schema.Question)
    @filter_queryset_for(form_schema.Form)
    @filter_queryset_for(form_schema.Option)
    @filter_queryset_for(form_schema.DynamicOption)
    def filter_queryset_public(self, node, queryset, info):
        # this is blueprint data which is uncritical and can be exposed publicly
        return queryset

    @filter_queryset_for(form_schema.Document)
    @filter_queryset_for(historical_form_schema.HistoricalDocument)
    def filter_queryset_for_document(self, node, queryset, info):
        instance_ids = self._all_visible_instances(info)

        return queryset.filter(
            # document is accessible through CAMAC ACLs
            Q(family__case__family__instance__pk__in=instance_ids)
            | Q(family__work_item__case__family__instance__pk__in=instance_ids)
            # OR dashboard documents
            | Q(form_id=DASHBOARD_FORM_SLUG)
            # OR unlinked table documents created by the requesting user
            | Q(
                family__case__instance__isnull=True,
                family=F("pk"),
                created_by_user=info.context.user.username,
            )
        )

    @filter_queryset_for(form_schema.Answer)
    @filter_queryset_for(historical_form_schema.HistoricalAnswer)
    def filter_queryset_for_answer(self, node, queryset, info):
        # return all answers, since answers can only be accessed via documents
        # which are already properly protected
        return queryset

    @filter_queryset_for(workflow_schema.Case)
    def filter_queryset_for_case(self, node, queryset, info):
        order_by = order(self.request)

        queryset = queryset.filter(
            family__instance__pk__in=self._all_visible_instances(info)
        )

        return queryset.order_by(*order_by) if order_by else queryset

    @filter_queryset_for(workflow_schema.WorkItem)
    def filter_queryset_for_work_items(self, node, queryset, info):
        return queryset.filter(
            case__family__instance__pk__in=self._all_visible_instances(info)
        )

    def _all_visible_instances(self, info):
        """Fetch visible camac instances and cache the result.

        Take user's group from a custom HTTP header named `X-CAMAC-GROUP` or use
        default group  to retrieve all Camac instance IDs that are accessible.

        Return a list of instance identifiers.
        """
        result = getattr(info.context, "_visibility_instances_cache", None)
        if result is not None:  # pragma: no cover
            return result

        filtered = CalumaInstanceFilterSet(
            data=filters(self.request),
            queryset=self.get_queryset(),
            request=self.request,
        )

        instance_ids = filtered.qs.values_list("pk", flat=True)

        setattr(info.context, "_visibility_instances_cache", instance_ids)
        return instance_ids


class CustomVisibilitySZ(CustomVisibility):
    """Custom visibility for Kanton Schwyz.

    Form visibility rules are defined in the form meta.
    Possible configurations are shown below.
    Note: If no visibility is configured for a form,
    the form is visible by default.

    Visible for internal roles:
    "meta": {
        "visibility": {
            "type": "internal"
        }
    },

    Publicly visible:
    "meta": {
        "visibility": {
            "type": "public"
        }
    },

    Visible for certain roles and services:
    * Request user needs to belong to one of the
    specified roles OR services:
    "meta": {
        "visibility": {
            "type": "specific",
            "visibleFor": {
                "roles": [3],
                "services": [6]
            }
        }
    },

    * Request user needs to belong to one of the specified
    roles, service is not relevant:
    "meta": {
        "visibility": {
            "type": "specific",
            "visibleFor": {
                "roles": [3, 4]
            }
        }
    },

    * Request user needs to belong to one of the specified
    services, role is not relevant:
    "meta": {
        "visibility": {
            "type": "specific",
            "visibleFor": {
                "services": [6, 7]
            }
        }
    },

    TODO: More generic future design
    To make the configurable visibility / permission feature
    usable for the other cantons, a few things need to be adapted:

    * Use "permissions" instead of "visibility" on form meta:

        "meta": {
            "createPermission": {
                "type": "public"
            }
        },

    * Read from and check permission config in instance serializer
        (create method), throw exception if permission is missing

    * Filter forms based on create permission in frontend
    """

    @filter_queryset_for(form_schema.Form)
    def filter_queryset_for_form(self, node, queryset, info):
        camac_role = self.request.group.role.role_id
        camac_service = self.request.group.service_id

        default_filter = Q(meta__visibility__isnull=True)

        # public filter: visible to all
        public_filter = Q(meta__visibility__type="public")

        # internal filter: excludes configured public roles
        public_roles = [
            role.role_id
            for role in Role.objects.filter(
                name__in=settings.APPLICATION.get("PUBLIC_ROLES", [])
            )
        ]
        internal_forms = (
            [form.slug for form in queryset.filter(meta__visibility__type="internal")]
            if camac_role not in public_roles
            else []
        )
        internal_filter = Q(slug__in=internal_forms)

        # specific filter: visible to specific roles or services
        specific_filter = Q(
            meta__visibility__type="specific",
            meta__visibility__visibleFor__roles__contains=[camac_role],
        ) | Q(
            meta__visibility__type="specific",
            meta__visibility__visibleFor__services__contains=[camac_service],
        )

        return queryset.filter(
            default_filter | public_filter | internal_filter | specific_filter
        )

    @filter_queryset_for(form_schema.Question)
    @filter_queryset_for(form_schema.Option)
    @filter_queryset_for(form_schema.DynamicOption)
    def filter_queryset_public(self, node, queryset, info):
        # this is blueprint data which is uncritical and can be exposed publicly
        return queryset

    @filter_queryset_for(workflow_schema.WorkItem)
    def filter_queryset_for_work_items(self, node, queryset, info):
        visibility_config = settings.APPLICATION.get("INTER_SERVICE_GROUP_VISIBILITIES")

        service = self.request.group.service

        queryset = queryset.filter(
            case__family__instance__pk__in=self._all_visible_instances(info)
        )

        if not service:
            return queryset

        # If the work-item task is "inquiry", check if the current service
        # is either in the controlling_groups or addressed_groups. If not, then
        # check if the current service is permitted to see the work-item
        # according to its service_group
        return queryset.filter(
            Case(
                When(
                    Q(task_id=settings.DISTRIBUTION["INQUIRY_TASK"])
                    & ~Q(addressed_groups__contains=[service.pk])
                    & ~Q(controlling_groups__contains=[service.pk]),
                    then=Exists(
                        Service.objects.filter(
                            # Use element = ANY(array) operator to check if
                            # element is present in ArrayField, which requires
                            # lhs and rhs of expression to be of same type
                            pk=Func(
                                Cast(
                                    OuterRef("addressed_groups"),
                                    output_field=ArrayField(IntegerField()),
                                ),
                                function="ANY",
                            ),
                            service_group__pk__in=visibility_config.get(
                                service.service_group_id, []
                            ),
                        )
                    ),
                ),
                default=True,
            ),
        )
