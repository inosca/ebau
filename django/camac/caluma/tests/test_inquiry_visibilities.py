from itertools import chain

import pytest
from caluma.caluma_core.relay import extract_global_id
from caluma.caluma_workflow.models import WorkItem

from camac.caluma.extensions.visibilities import CustomVisibility, CustomVisibilityBE
from camac.caluma.utils import visible_inquiries_expression
from camac.user.models import Service


@pytest.mark.parametrize("role__name", ["Municipality"])
def test_inquiry_visibility_be(
    db,
    caluma_admin_schema_executor,
    be_distribution_settings,
    active_inquiry_factory,
    role,
    mocker,
    service,
    be_instance,
    service_factory,
    settings,
    gql,
):
    settings.APPLICATION_NAME = "kt_bern"
    mocker.patch(
        "caluma.caluma_core.types.Node.visibility_classes", [CustomVisibilityBE]
    )

    other_service = service_factory()
    other_subservice = service_factory(service_parent=other_service)

    visible_inquiries = [
        active_inquiry_factory(controlling_service=c, addressed_service=a)
        for c, a in [
            (other_service, service),
            (service, other_service),
            (service_factory(), other_service),
        ]
    ]

    # An inquiry addressed to a subservice in which we are not involved should
    # not be visible to us
    invisible_inquiry = active_inquiry_factory(
        controlling_service=other_service, addressed_service=other_subservice
    )

    result = caluma_admin_schema_executor(
        gql("work-items-for-task"),
        variables={"task": be_distribution_settings["INQUIRY_TASK"]},
    )

    assert not result.errors

    ids = set(
        [
            extract_global_id(edge["node"]["id"])
            for edge in result.data["allWorkItems"]["edges"]
        ]
    )

    visible_ids = {str(i.pk) for i in visible_inquiries}

    assert len(ids) == 3
    assert visible_ids == ids
    assert str(invisible_inquiry.pk) not in ids


@pytest.mark.parametrize(
    "role__name,expected_count",
    [("service", 0), ("subservice", 3)],
)
def test_inquiry_visibility_gr(
    db,
    caluma_admin_schema_executor,
    gr_distribution_settings,
    active_inquiry_factory,
    role,
    mocker,
    service,
    gr_instance,
    service_factory,
    settings,
    expected_count,
    gql,
):
    settings.APPLICATION_NAME = "kt_gr"
    mocker.patch("caluma.caluma_core.types.Node.visibility_classes", [CustomVisibility])

    other_service = service_factory()
    other_subservice = service_factory(service_parent=other_service)

    [
        active_inquiry_factory(controlling_service=c, addressed_service=a)
        for c, a in [
            (other_service, service),
            (service, other_service),
            (other_service, other_subservice),
            (service_factory(), other_service),
        ]
    ]

    result = caluma_admin_schema_executor(
        gql("work-items-for-task"),
        variables={"task": gr_distribution_settings["INQUIRY_TASK"]},
    )

    assert not result.errors

    ids = set(
        [
            extract_global_id(edge["node"]["id"])
            for edge in result.data["allWorkItems"]["edges"]
        ]
    )

    assert len(ids) == expected_count


@pytest.fixture
def so_inquiries(
    service_factory, active_inquiry_factory, user_group_factory, admin_user, gr_instance
):
    root_service = service_factory(name="root_service")
    service_1 = service_factory(name="service_1")
    service_2 = service_factory(name="service_2")

    service_1_subservice_1 = service_factory(
        name="service_1_subservice_1", service_parent=service_1
    )
    service_1_subservice_2 = service_factory(
        name="service_1_subservice_2", service_parent=service_1
    )
    service_2_subservice_1 = service_factory(
        name="service_2_subservice_1", service_parent=service_2
    )

    [
        active_inquiry_factory(controlling_service=c, addressed_service=a)
        for c, a in [
            (root_service, service_1),
            (root_service, service_2),
            (service_1, service_1_subservice_1),
            (service_1, service_1_subservice_2),
            (service_2, service_2_subservice_1),
        ]
    ]

    def select_service(name):
        user_group = user_group_factory(
            user=admin_user, group__service=Service.objects.get(name=name)
        )

        return user_group.group

    return select_service


@pytest.mark.parametrize("service_name", ["service_1", "service_1_subservice_1"])
def test_inquiry_visibility_so(
    db,
    caluma_admin_schema_executor_for_group,
    so_distribution_settings,
    gr_instance,
    so_inquiries,
    service_name,
    mocker,
    settings,
    gql,
):
    settings.APPLICATION_NAME = "kt_so"
    mocker.patch("caluma.caluma_core.types.Node.visibility_classes", [CustomVisibility])
    mocker.patch(
        "camac.caluma.extensions.visibilities.CustomVisibility._all_visible_instances",
        return_value=[gr_instance.pk],
    )

    group = so_inquiries(service_name)

    result = caluma_admin_schema_executor_for_group(group.pk)(
        gql("work-items-for-task"),
        variables={"task": so_distribution_settings["INQUIRY_TASK"]},
    )

    assert not result.errors

    visible_service_ids = set(
        chain(
            *[
                [int(id) for id in edge["node"]["addressedGroups"]]
                for edge in result.data["allWorkItems"]["edges"]
            ]
        )
    )

    expected_service_ids = set(
        Service.objects.filter(
            name__in=[
                "service_1",
                "service_2",
                "service_1_subservice_1",
                "service_1_subservice_2",
            ]
        ).values_list("pk", flat=True)
    )

    assert expected_service_ids == visible_service_ids


def test_inquiry_visibility_ag(
    db,
    ag_distribution_settings,
    case_factory,
    group,
    service_factory,
    settings,
    snapshot,
    task_factory,
    work_item_factory,
):
    settings.APPLICATION_NAME = "kt_ag"
    task = task_factory(pk=ag_distribution_settings["INQUIRY_TASK"])

    authority = service_factory(service_group__name="municipality")
    authority_subservice = service_factory(
        service_group__name="municipality", service_parent=authority
    )
    municipality = service_factory(service_group__name="municipality")
    municipality_subservice = service_factory(
        service_group__name="municipality", service_parent=municipality
    )

    afb = service_factory(service_group__name="service-afb")
    afb_subservice = service_factory(
        service_group__name="service-afb", service_parent=afb
    )

    canton = service_factory(service_group__name="service-cantonal")
    canton_subservice = service_factory(
        service_group__name="service-cantonal", service_parent=canton
    )

    external = service_factory(service_group__name="service-external")
    external_subservice = service_factory(
        service_group__name="service-external", service_parent=external
    )

    _case = case_factory()

    for name, controlling, addressed in [
        ("authority-to-municipality", authority, municipality),
        ("authority-to-subservice", authority, authority_subservice),
        ("authority-to-canton", authority, canton),
        ("authority-to-afb", authority, afb),
        ("authority-to-external", authority, external),
        ("municipality-to-subservice", municipality, municipality_subservice),
        ("afb-to-canton", afb, canton),
        ("afb-to-external", afb, external),
        ("afb-to-subservice", afb, afb_subservice),
        ("canton-to-subservice", canton, canton_subservice),
        ("external-to-subservice", external, external_subservice),
    ]:
        work_item_factory(
            task=task,
            case=_case,
            name=name,
            addressed_groups=[str(addressed.pk)],
            controlling_groups=[str(controlling.pk)],
        )

    services = dict(
        authority=authority,
        authority_subservice=authority_subservice,
        municipality=municipality,
        municipality_subservice=municipality_subservice,
        afb=afb,
        afb_subservice=afb_subservice,
        canton=canton,
        canton_subservice=canton_subservice,
        external=external,
        external_subservice=external_subservice,
    )

    for name, service in services.items():
        group.service = service
        inquiries = WorkItem.objects.filter(task_id=task, case=_case).filter(
            visible_inquiries_expression(group)
        )

        assert sorted(set(inquiries.values_list("name__de", flat=True))) == snapshot(
            name=name
        )
