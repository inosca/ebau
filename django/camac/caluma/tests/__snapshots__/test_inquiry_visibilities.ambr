# serializer version: 1
# name: test_inquiry_visibility_ag[afb]
  list([
    'afb-to-canton',
    'afb-to-external',
    'afb-to-subservice',
    'authority-to-afb',
    'authority-to-canton',
    'authority-to-external',
    'canton-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[afb_subservice]
  list([
    'afb-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[authority]
  list([
    'authority-to-afb',
    'authority-to-canton',
    'authority-to-external',
    'authority-to-municipality',
    'authority-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[authority_subservice]
  list([
    'authority-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[canton]
  list([
    'afb-to-canton',
    'afb-to-external',
    'authority-to-afb',
    'authority-to-canton',
    'authority-to-external',
    'canton-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[canton_subservice]
  list([
    'canton-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[external]
  list([
    'afb-to-external',
    'authority-to-external',
    'external-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[external_subservice]
  list([
    'external-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[municipality]
  list([
    'authority-to-municipality',
    'municipality-to-subservice',
  ])
# ---
# name: test_inquiry_visibility_ag[municipality_subservice]
  list([
    'municipality-to-subservice',
  ])
# ---
